### ISSUE TITLE
Timeout on publish

### ISSUE THREAD

**Issue by dushyantbangal** [mqttjs/MQTT.js#816]
I have a case in which I want to store the messages to a local database in case publish fails, and then retry after some time.

However, the publish (qos:1) callback does not get called at all when the network is not present.
I know it sends the messages after the connection is up again, but it will fail if the script restarts (due to power loss or something else)

Is this expected behavior? Is there any way I can achieve what I want to?

**Comment by zlatinejc** [comment]
look in yhe store.js section to implement your persistance layer

**Comment by dushyantbangal** [comment]
If you mean https://github.com/mqttjs/MQTT.js#mqttstoreoptions then yes, but that was kinda my last option. After I posted this question, I looked into it a bi deeper and I think mqtt-nedbb-store should help me do what I need to.

**Comment by matiasdecarli** [comment]
Hi @dushyantbangal . I believe you have the same issue I had a couple of days ago. Take, a look

https://github.com/aws/aws-iot-device-sdk-js/issues/226

tl;dr, you can set the timeout on the `keepalive` option in seconds

**Comment by dushyantbangal** [comment]
@matiasdecarli do you mean the callback will be fired after 10 secs if I set `keepalive:10` and I haven't received `PUBACK` in that time?
Because my main issue is callback not being called.

**Comment by matiasdecarli** [comment]
I'm not 100% sure if the callback is going to get called, but the library is going to try to reconnect after 10 seconds

**Comment by dushyantbangal** [comment]
Yes, the reconnection is not an issue for me, it's being handled automatically by the library. Callback with error is what is needed, which I think might be a bit complicated to implement.

Anyways, this issue is mostly solved by using store.
Package maintainers can take a call whether to close this issue or not.

**Comment by malishahi** [comment]
I think the solution is to check if a client is connected, before calling publish. If client is not connected you don't have to publish, if you wanna rely on its callback, since callback does not get fired at all when client is not connected due to network problems, or broker being down.

**Comment by dushyantbangal** [comment]
@malishahi Sometimes the network will go down and the client will still say its connected, because it hasn't performed its keepAlive yet.

**Comment by vishnureddy17** [comment]
> Anyways, this issue is mostly solved by using store. Package maintainers can take a call whether to close this issue or not.
Closing issue.
